### ISSUE TITLE
What are causes for `error: archive is not a valid package` when trying to publish a package?

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#330]
I have tried to publish `v0.0.1+1` of [Geometry Central's build2 package](https://github.com/build2-packaging/geometry-central) and got the following log with a trailing error message:

```
publishing:
  to:      https://cppget.org
  as:      lyrahgames <lyrahgames@mailbox.org>

  package: libgeometrycentral
  version: 0.0.1+1
  project: geometry-central
  section: alpha
  control: https://github.com/build2-packaging/geometry-central.git

  package: libgeometrycentral-tests
  version: 0.0.1+1
  project: geometry-central
  section: alpha
  control: https://github.com/build2-packaging/geometry-central.git
warning: publishing using staged build2 toolchain
continue? [y/n] y
pushing branch build2-control
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 8 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 345 bytes | 345.00 KiB/s, done.
Total 3 (delta 1), reused 0 (delta 0), pack-reused 0
submitting libgeometrycentral-0.0.1+1.tar.gz
############################################################################################################### 100.0%
error: archive is not a valid package (run bpkg pkg-verify for details)
  info: reference: 43ee2d0d4d9b
```

I am using the up-to-date staged build2 toolchain and assume that the package might miss something or contain an error. To locate its origin, can you tell me, what could possibly cause a package to be invalid? I would also run `bpkg pkg-verify` as stated myself. But the `*.tar.gz` is removed after the operation and I do not know how to generate it.

**Comment by karen-arutyunov** [comment]
Geven geometry-central/libgeometrycentral/ is bdep-initialized, you can generate the archive with the following command:

```
$ b dist: geometry-central/libgeometrycentral/ config.dist.root=. config.dist.archives=tar.gz
```

To run `bpkg-pkg-verify` use the toolchain which is used by the repository you upload the archive to (currently 0.16.0 for cppget.org) :

```
$ bpkg pkg-verify libgeometrycentral-0.0.1+1.tar.gz
error: unable to satisfy constraint (build2 >= 0.17.0-) for package libgeometrycentral-0.0.1+1.tar.gz
  info: available build2 version is 0.16.0
```

**Comment by lyrahgames** [comment]
Thank you very much for the fast answer and the instructions!
Up to now, I thought that `cppget.org` would use the staged toolchain when requiring it and, as a consequence, would put packages built with it into the alpha section. For now, this solves my issue and I will wait with publishing until the next version release and `cppget.org` update.

**Comment by boris-kolpackov** [comment]
Ok, we've improved the diagnostics in `bdep-publish` result and (once the staged toolchain is releaves) you should get:

```
error: package archive is not valid
  info: unable to satisfy constraint (build2 >= 0.17.0-) for package libhello-1.0.0.tar.gz
  info: available build2 version is 0.16.0
```

> Up to now, I thought that `cppget.org` would use the staged toolchain when requiring it and, as a consequence, would put packages built with it into the alpha section. 

No, `cppget.org` only runs one version of the toolchain. Running different versions for different sections would kick the complexity to the next level so this is unlikely to happen.

>  For now, this solves my issue and I will wait with publishing until the next version release and `cppget.org` update.

I've documented the available alternatives in this situation here:

https://build2.org/stage/build2-toolchain/doc/build2-toolchain-packaging.xhtml#faq-publish-stage

I think in your case option (2) could work.
